internal/middleware/modelid.go

```diff
@@ -0,0 +1,74 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	modelIDCtxKey ctxKey = "model-id-context-key"
+
+	modelIDTraceKey = "authorization_model_id"
+	modelIDHeader   = "authorization-model-id"
+)
+
+func ModelIDFromContext(ctx context.Context) (string, bool) {
+	modelID, ok := ctx.Value(modelIDCtxKey).(string)
+	return modelID, ok
+}
+
+type hasGetAuthorizationModelID interface {
+	GetStoreId() string
+	GetAuthorizationModelId() string
+}
+
+// NewModelIDInterceptor must come after the trace and storeID interceptors
+func NewModelIDInterceptor(ds storage.AuthorizationModelReadBackend) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if r, ok := req.(hasGetAuthorizationModelID); ok {
+			modelID := r.GetAuthorizationModelId()
+
+			if modelID == "" {
+				storeID := r.GetStoreId()
+				if storeID == "" {
+					return nil, errors.New("no store id")
+				}
```

Comment:
this code is not covered by tests. Can it ever happen? We have non-empty validation at the API layer, e.g. https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L803